framework/aster-frame/src/sync/wait.rs

```diff
@@ -111,28 +113,54 @@ impl WaitQueue {
 
     /// Wake up one waiting thread.
     pub fn wake_one(&self) {
-        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
+        // Fast path
+        if self.is_empty() {
+            return;
+        }
+
+        loop {
+            let mut wakers = self.wakers.lock_irq_disabled();
+            let Some(waker) = wakers.pop_front() else {
+                break;
+            };
+            self.count.fetch_sub(1, Ordering::Relaxed);
+
             // Avoid holding lock when calling `wake_up`
+            drop(wakers);
             if waker.wake_up() {
-                return;
+                break;
             }
         }
     }
 
     /// Wake up all waiting threads.
     pub fn wake_all(&self) {
-        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
+        // Fast path
+        if self.is_empty() {
+            return;
+        }
+
+        loop {
+            let mut wakers = self.wakers.lock_irq_disabled();
+            let Some(waker) = wakers.pop_front() else {
+                break;
+            };
+            self.count.fetch_sub(1, Ordering::Relaxed);
+
             // Avoid holding lock when calling `wake_up`
+            drop(wakers);
             waker.wake_up();
         }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.wakers.lock_irq_disabled().is_empty()
+        self.count.load(Ordering::Relaxed) == 0
```

Comment:
I think we should use the `Release-Acquire` memory ordering.

https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering

Reply:
No, I don't think this is correct. If it is correct, why should I write `self.count.fetch_add(0, Ordering::Release) == 0`, which looks pretty wired at first glance?

The exact point is that you need to make sure that **memory writes before `is_empty()` cannot be reordered after the emptiness check**, which means you need to have a `Release` operation on `is_empty()`, not `Acquire`.

Quoted from the cppreference you've attached:
 - A load operation with this memory order performs the **acquire** operation on the affected memory location: **no reads or writes in the current thread can be reordered before this load**. 
 - A store operation with this memory order performs the **release** operation: **no reads or writes in the current thread can be reordered after this store**.

If we use `fetch_add(0, Ordering::Release)` to load the counter, the compiler will know that we need a `Release` operation here, so it will insert a `mfence` instruction before the load (check the result of "show assembly" [in the Rust playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=ac7c8af3f2f3c9c24f7bc9a4efe910e7)):
```asm
playground::load_acquire:
	movq	playground::COUNTER0(%rip), %rax
	retq

playground::load_release:
	mfence
	movq	playground::COUNTER0(%rip), %rax
	retq
```
The `mfence` instruction is exactly what we need, and it is exactly what the [`smp_mb`](https://github.com/torvalds/linux/blob/cf87f46fd34d6c19283d9625a7822f20d90b64a4/arch/x86/include/asm/barrier.h#L22) function in the [Linux kernel implementation](https://github.com/torvalds/linux/blob/f4345f05c0dfc73c617e66f3b809edb8ddd41075/include/linux/wait.h#L95-L124) does. Without the `mfence` instruction, the stores (which update the result of the `cond()` check) can be reordered after the emptiness check, and we will miss a wake-up event in these scenarios.

To be honest, a wrong memory order may improve performance (by avoiding the overhead of the `mfence' statement). Meanwhile, this problem will only become apparent when we run heavy workloads on multiple processors. So we can leave a TODO here and be happy for a really long time, but it's still problematic, isn't it?

I will revisit the memory order used in the `WaitQueue`/`Waker`/`Waiter` when I have some spare time, so you don't need to file another PR to fix this problem. (But of course you can if you want.)